resources/assets/lib/click-menu.ts

```diff
@@ -0,0 +1,135 @@
+/**
+ *    Copyright (c) ppy Pty Ltd <contact@ppy.sh>.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Fade from 'fade';
+
+export default class ClickMenu {
+  private current: string | null | undefined = null;
+
+  constructor() {
+    $(document).on('click', '.js-click-menu[data-click-menu-target]', this.toggle);
+    $(document).on('click', this.hide);
+    document.addEventListener('turbolinks:load', this.restoreSaved);
+    document.addEventListener('turbolinks:before-cache', this.saveCurrent);
+  }
+
+  closestMenuId(child: Element | null | undefined) {
+    if (child != null) {
+      return $(child).parents('[data-click-menu-id]').attr('data-click-menu-id');
+    }
+  }
+
+  hide = (e: JQuery.ClickEvent) => {
+    if (e.button !== 0) return;
+    if (osu.popupShowing()) return;
+    if (this.current == null) return;
+    if ($(e.target).closest('.js-click-menu').length > 0) return;
+
+    this.show();
+  }
+
+  menuLink(id: string | null | undefined) {
+    return document.querySelector(`.js-click-menu[data-click-menu-target${id == null ? '' : `='${id}'`}]`);
+  }
+
+  restoreSaved = () => {
+    this.show(document.body.dataset.clickMenuCurrent);
+  }
+
+  saveCurrent = () => {
+    if (this.current == null) {
+      delete document.body.dataset.clickMenuCurrent;
+    } else {
+      document.body.dataset.clickMenuCurrent = this.current;
+    }
+  }
+
+  show = (target?: string | null | undefined) => {
+    this.current = target;
+
+    const tree = this.tree();
+    const menus = Array.from(document.querySelectorAll('.js-click-menu[data-click-menu-id]'));
```

Comment:
apparently adding `dom.iterable` to `lib` in `tsconfig.json` removes the need to use `Array.from`